src/scope.rs

```diff
@@ -138,23 +137,72 @@ pub fn analyze<'a>(files: &'a Map<PathBuf, Parsed>, config: &ConfigFile, emitter
                         .what(usage.name.clone())
                         .emit(emitter);
                 }
-                Some((Found::Indirect, _)) => {
+                Some((Found::Indirect, item)) => {
+                    let imported_through = parsed.imports
+                        .keys()
+                        .find(|imported_file| {
+                            get_cached_scope(imported_file, &scopes)
+                                .search(&usage.name)
+                                .is_some()
+                        })
+                        .unwrap_or_else(|| unreachable!());
+
+                    used_dependencies.insert(imported_through);
+
                     usage.location.in_file(&parsed.original_path)
                         .lint(Lint::IndirectImports, format!("Indirectly imported: {}", usage.name))
                         .what(usage.name.clone())
+                        .note(format!(
+                            "Indirectly imported through {}",
+                            files[imported_through].original_path.display()
+                        ))
+                        .note(format!(
+                            "Consider directly importing {}",
+                            files[item.origin].original_path.display()
+                        ))
                         .emit(emitter);
                 }
                 _ => ()
             }
             if let Some((_, item)) = search_result {
+                used_dependencies.insert(item.origin);
+
                 if usage.name != item.name {
                     usage.location.in_file(&parsed.original_path)
                         .lint(Lint::InvalidLetterCasing, "Function name differs between usage and definition")
-                        .note(format!("Check whether the letter casing is the same"))
+                        .note("Check whether the letter casing is the same")
                         .emit(emitter);
                 }
             }
         }
+
+        // TODO perhaps we can move this check out of scope
+        // analysis to its own module? That would require
+        // scope analysis to save some info.
+        for (imported_file, import) in &parsed.imports {
+            if !used_dependencies.contains(&**imported_file) {
+                if files[imported_file]
+                        .definitions
+                        .iter()
+                        .filter(|def| !def.name.starts_with("!")) // skip pseudoitems
```

Comment:
I don't get this? What is a pseudoitem

Reply:
Pseudoitems are things not really defined in the source code, but something we inject in analysis. Eg. there's an pseudo-item ‘This file “defines” usage of strict mode’. Its name starts with `!` so it doesn't clash with real functions.

Reply:
Wrapped this iterator chain in a function, so it's more obvious now.